File: packages/npm/locale.js
/** Korean primary language => ko. Every other language, including C/POSIX, => en. */
export function resolveLanguage(value) {
  return typeof value === 'string' && /^ko(?:[-_.@:]|$)/i.test(value.trim()) ? 'ko' : 'en';
}
export function detectLanguage({ language, browserLanguage, env, systemLanguage } = {}) {
  if (language && language !== 'auto') return resolveLanguage(language);
  if (env) {
    for (const key of ['TESSEMBLY_LANG', 'LC_ALL', 'LC_MESSAGES', 'LANG', 'LANGUAGE']) {
      if (env[key] && env[key] !== 'auto') return resolveLanguage(env[key].split(':')[0]);
    }
  }
  return resolveLanguage(browserLanguage || systemLanguage || 'en');
}
const messages = {
  INVALID_ARGUMENTS: ['Invalid arguments.', '인수가 올바르지 않습니다.'],
  PROFILE_REQUIRED: ['Specify the semantic profile.', '의미 프로필을 지정하세요.'],
  UNSUPPORTED_PROFILE: ['This semantic profile is not supported.', '지원하지 않는 의미 프로필입니다.'],
  MIGRATION_REQUIRES_METADATA_HANDLER: ['Unknown metadata requires a profile-aware migration handler.', '알 수 없는 메타데이터에는 프로필을 아는 별도 이관 처리가 필요합니다.'],
  CONFIG_CONFLICT: ['Document and host settings conflict.', '문서 설정과 호스트 설정이 충돌합니다.'],
  INVALID_UTF8: ['Input must be valid UTF-8.', '올바른 UTF-8 입력이 필요합니다.'],
  INPUT_LIMIT: ['Input exceeds the supported size.', '입력 크기가 지원 한도를 초과했습니다.'],
  READ_FAILED: ['Could not read the file.', '파일을 읽을 수 없습니다.'],
  WRITE_FAILED: ['Could not write the file.', '파일을 쓸 수 없습니다.'],
  OPAQUE_METADATA_WOULD_BE_LOST: ['Text export would discard opaque metadata.', '텍스트 변환 시 불투명 메타데이터가 유실됩니다.'],
  WASM_LOAD_FAILED: ['Could not load the bundled WebAssembly module.', '포함된 WebAssembly 모듈을 불러올 수 없습니다.'],
  WASM_ABI_MISMATCH: ['The WebAssembly ABI does not match this package.', 'WebAssembly ABI가 패키지와 일치하지 않습니다.'],
  WASM_TRAP: ['WebAssembly processing stopped; no complete result is available.', 'WebAssembly 처리가 중단되어 완전한 결과가 없습니다.'],
  WASM_INSTANCE_FAILED: ['This instance stopped after a trap; create a new instance.', '트랩 이후 중단된 인스턴스입니다. 새 인스턴스를 만드세요.'],
  WASM_SIZE_LIMIT: ['The Wasm asset exceeds the supported limit.', 'Wasm 자산이 크기 한도를 초과했습니다.'],
  SHARED_INPUT_UNSUPPORTED: ['Copy shared input into a private buffer first.', '공유 입력을 먼저 독립 버퍼로 복사하세요.'],
  REGULAR_FILE_REQUIRED: ['Input must be a regular file.', '입력은 일반 파일이어야 합니다.'],
  EEXIST: ['Output already exists; choose a new path.', '출력 파일이 이미 존재합니다. 새 경로를 지정하세요.'],
  DISPOSED: ['This Tessembly instance has been disposed.', '종료된 테섬블리 인스턴스입니다.']
};
export function errorMessage(code, language = 'en') {
  const ko = resolveLanguage(language) === 'ko';
  if (Object.hasOwn(messages, code)) return messages[code][ko ? 1 : 0];
  if (code.startsWith('UNSUPPORTED') || code.endsWith('REQUIRES_HOST')) return ko ? '이 기능 또는 상태는 호스트의 지원이 필요합니다.' : 'This feature or state requires host support.';
  if (code.endsWith('LIMIT') || code === 'INCOMPLETE') return ko ? '처리 한도를 초과했습니다. 불가능 판정이 아닙니다.' : 'A processing limit was reached; this does not prove impossibility.';
  if (code.startsWith('DUPLICATE')) return ko ? '중복된 선언 또는 식별자입니다.' : 'A declaration or identifier is duplicated.';
  if (code.startsWith('EXPECTED') || code.startsWith('INVALID') || code.startsWith('EMPTY') || code.startsWith('UNKNOWN')) return ko ? '입력이 문법 또는 구조 계약에 맞지 않습니다.' : 'Input does not meet the syntax or structure contract.';
  return ko ? '처리할 수 없는 입력입니다. 오류 코드와 위치를 확인하세요.' : 'Input could not be processed. Check the error code and byte span.';
}
export class TessemblyError extends Error {
  constructor(code, { language = 'en', start = 0, end = 0, cause } = {}) {
    super(`${errorMessage(code, language)} [${code}]`, { cause });
    this.name = 'TessemblyError'; this.code = code; this.start = start; this.end = end;
    this.language = resolveLanguage(language);
  }
}

File: crates/tessembly-cli/src/locale.rs
//! Presentation-only locale settings. Machine codes and format data stay stable.
use std::env;
use tessembly_core::{Error, Result};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Language {
    En,
    Ko,
}

pub fn resolve(value: &str) -> Language {
    let primary = value
        .trim()
        .split(['-', '_', '.', '@', ':'])
        .next()
        .unwrap_or("");
    if primary.eq_ignore_ascii_case("ko") {
        Language::Ko
    } else {
        Language::En
    }
}
fn from_settings(mut get: impl FnMut(&str) -> Option<String>) -> Language {
    for key in [
        "TESSEMBLY_LANG",
        "LC_ALL",
        "LC_MESSAGES",
        "LANG",
        "LANGUAGE",
    ] {
        if let Some(value) = get(key).filter(|v| !v.trim().is_empty() && v != "auto") {
            return resolve(&value);
        }
    }
    Language::En
}
pub fn detect() -> Language {
    from_settings(|key| env::var(key).ok())
}
pub fn arguments(args: &mut Vec<String>) -> Result<Language> {
    let Some(index) = args.iter().position(|v| v == "--lang") else {
        return Ok(detect());
    };
    let value = args
        .get(index + 1)
        .ok_or_else(|| Error::new("INVALID_ARGUMENTS"))?
        .clone();
    if !matches!(value.as_str(), "en" | "ko" | "auto") {
        return Err(Error::new("INVALID_ARGUMENTS"));
    }
    args.drain(index..=index + 1);
    if args.iter().any(|v| v == "--lang") {
        return Err(Error::new("INVALID_ARGUMENTS"));
    }
    Ok(if value == "auto" {
        detect()
    } else {
        resolve(&value)
    })
}
pub fn help(language: Language) -> &'static str {
    match language {
        Language::Ko => include_str!("../../../docs/HELP.ko.md"),
        Language::En => include_str!("../../../docs/HELP.en.md"),
    }
}
pub fn message(code: &str, language: Language) -> &'static str {
    let pair = match code {
        "INVALID_ARGUMENTS" => ("Invalid arguments.", "인수가 올바르지 않습니다."),
        "PROFILE_REQUIRED" => ("Specify a semantic profile.", "의미 프로필을 지정하세요."),
        "UNSUPPORTED_PROFILE" => (
            "Unsupported semantic profile.",
            "지원하지 않는 의미 프로필입니다.",
        ),
        "MIGRATION_REQUIRES_METADATA_HANDLER" => (
            "Unknown metadata needs a profile-aware migration handler.",
            "알 수 없는 메타데이터에는 프로필을 아는 별도 이관 처리가 필요합니다.",
        ),
        "CONFIG_CONFLICT" => (
            "Document and host settings conflict.",
            "문서와 호스트 설정이 충돌합니다.",
        ),
        "READ_FAILED" => ("Could not read the file.", "파일을 읽을 수 없습니다."),
        "WRITE_FAILED" => ("Could not write the file.", "파일을 쓸 수 없습니다."),
        "OPAQUE_METADATA_WOULD_BE_LOST" => (
            "Text export would lose opaque metadata.",
            "텍스트 변환 시 불투명 메타데이터가 유실됩니다.",
        ),
        _ if code.ends_with("LIMIT") || code == "INCOMPLETE" => (
            "Processing limit reached; not an impossibility proof.",
            "처리 한도를 초과했습니다. 불가능 판정이 아닙니다.",
        ),
        _ if code.starts_with("UNSUPPORTED") || code.ends_with("REQUIRES_HOST") => (
            "The host must support this feature or state.",
            "이 기능이나 상태는 호스트의 지원이 필요합니다.",
        ),
        _ => (
            "Input does not meet the format contract. Check the code and byte span.",
            "입력이 형식 계약에 맞지 않습니다. 오류 코드와 바이트 위치를 확인하세요.",
        ),
    };
    if language == Language::Ko {
        pair.1
    } else {
        pair.0
    }
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn primary_language_only() {
        for value in ["ko", "KO-kr", "ko_KR.UTF-8"] {
            assert_eq!(resolve(value), Language::Ko);
        }
        for value in ["en-US", "ja", "kok", "C", "POSIX", "en:ko", ""] {
            assert_eq!(resolve(value), Language::En);
        }
    }
    #[test]
    fn environment_precedence_is_not_secondary_language_search() {
        assert_eq!(
            from_settings(|k| match k {
                "LC_ALL" => Some("C".into()),
                "LANG" => Some("ko_KR".into()),
                _ => None,
            }),
            Language::En
        );
    }
}
